### PR TITLE
Added documentation on updated IAM behavior

### PR DESCRIPTION
**Pull request by renebrandel**
### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] Swift
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://docs.amplify.aws/)` 
            HTML: `<a href="https://docs.amplify.aws/">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx [hunk 1]
```diff
@@ -139,17 +139,25 @@ type Todo @model @auth(rules: [{ allow: public }]) {
...
+You can also override the authorization provider. In the example below, you can use an "Unauthenticated Role" from the Cognito identity pool for public access instead of an API Key.
 
 <BlockSwitcher>
 <Block name="Amplify CLI">
 
 When you run `amplify add auth`, the Amplify CLI generates scoped down IAM policies for the "Unauthenticated role" in Cognito identity pool automatically.
 
+```graphql
+# public authorization with provider override
+type Post @model @auth(rules: [{ allow: public, provider: iam }]) {
```
Comment (sobolk) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:151]:
```suggestion
type Post @model @auth(rules: [{ allow: public, provider: identityPool }]) {
```

should we do this now or later ?

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx [hunk 2]
```diff
@@ -297,17 +307,24 @@ type Todo @model @auth(rules: [{ allow: private }]) {
...
+You can also override the authorization provider. In the example below, you can use an "Authenticated Role" from the Cognito identity pool for granting access to signed-in users.
 
 <BlockSwitcher>
 <Block name="Amplify CLI">
 
 When you run `amplify add auth`, the Amplify CLI generates scoped down IAM policies for the "Authenticated role" in Cognito identity pool automatically.
 
+```graphql
+# public authorization with provider override
+type Post @model @auth(rules: [{ allow: private, provider: iam }]) {
```
Comment (sobolk) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:319]:
```suggestion
type Post @model @auth(rules: [{ allow: private, provider: identityPool }]) {
```

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx [hunk 3]
```diff
@@ -672,12 +692,11 @@ userPools:owner:owner
 
 ### Use IAM authorization within the AppSync console
 
-IAM-based `@auth` rules are scoped down to only work with Amplify-generated IAM roles. To access the GraphQL API with IAM authorization within your AppSync console, you need to explicitly allow list the IAM user's name.
 
 <BlockSwitcher>
 <Block name="Amplify CLI">
 
-Add the allow-listed IAM users by adding them to `amplify/backend/api/<your-api-name>/custom-roles.json`. (Create the `custom-roles.json` file if it doesn't exist). Append the `adminRoleNames` array with the IAM role or user names:
+IAM-based `@auth` rules are scoped down to only work with Amplify-generated IAM roles. To access the GraphQL API with IAM authorization within your AppSync console, you need to explicitly allow list the IAM user's name. Add the allow-listed IAM users by adding them to `amplify/backend/api/<your-api-name>/custom-roles.json`. (Create the `custom-roles.json` file if it doesn't exist). Append the `adminRoleNames` array with the IAM role or user names:
```
Comment (sobolk) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:699]: Does this paragraph make sense ?

All they need to do is to set `authorizationModes.iamConfig.enableIamAuthorizationMode = true` and attach appropriate IAM policy to any non-cognito IAM principal. 

If we're deprecating `adminRoleNames` we should write this section from scratch. 

Or perhaps delete this block and just mention in section below that any non-cognito IAM principal would work.

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx [hunk 4]
```diff
@@ -688,7 +707,7 @@ Add the allow-listed IAM users by adding them to `amplify/backend/api/<your-api-
 </Block>
 <Block name="AWS CDK">
 
-To grant an external AWS Resource or an IAM role access to this GraphQL API in CDK, you need to explicitly list the IAM role by adding them to `adminRoles` property.
+To grant an external AWS Resource or an IAM role access, **with the exception of Amazon Cognito identity pool roles**, to this GraphQL API in CDK, you need to enable IAM authorization mode via the `iamConfig` property of the CDK construct.
```
Comment (sobolk) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:710]:
```suggestion
To grant any IAM principal (AWS Resource, IAM role, IAM user, etc) access, **with the exception of Amazon Cognito identity pool roles**, to this GraphQL API in CDK, you need to enable IAM authorization mode via the `iamConfig` property of the CDK construct.
```

It's not only resources and roles, but also IAM users.

File: src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+1. Upgrade to the latest version of the Amplify GraphQL API CDK construct by running `npm install @aws-amplify/graphql-api-construct@latest`
+2. Set the `enableIamAuthorizationMode` to `true` under `iamConfig`.
+3. Move the existing values for `identityPoolId`, `authenticatedUserRole`, `unauthenticatedUserRole` from `iamConfig` to `identityPoolConfig`.
+4. Validate that all IAM roles has the appropriate permissions to access the GraphQL API. **If you have configured prior overly permissive policies, modify the policies accordingly.**
+5. Unset the values for `allowListedAdminRoles`. Your `iamConfig` should only have a `enableIamAuthorizationMode: true` value at this point.
+6. Deploy your GraphQL API with the updated configuration via `cdk deploy`.
+
+## Is this a breaking change?
+
+No. Your existing IAM authorization behavior will continue to work but we plan on deprecating the old behavior in the future, thus we highly recommend you to upgrade to the new behavior. If the following values under.
```
Comment (palpatim) [inline src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx:66]: > If the following values under.

Incomplete?

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx [hunk 6]
```diff
@@ -811,7 +833,7 @@ Once you grant a function access to the GraphQL API, it is required to redeploy
 </Block>
 
 <Block name="AWS CDK">
-To grant an external AWS Resource or an IAM role access to this GraphQL API in CDK, you need to explicitly list the IAM role by adding them to `adminRoles` property.
+To grant an external AWS Resource or an IAM role access, **with the exception of Amazon Cognito identity pool roles**, to this GraphQL API in CDK, you need to enable IAM authorization mode on the CDK construct.
```
Comment (sobolk) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:836]: similar comment here.

File: src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+## Who can take advantage of this improvement?
+
+Only if you use the [Amplify GraphQL API CDK construct](https://constructs.dev/packages/@aws-amplify/graphql-api-construct/v/1.6.0?lang=typescript), you'll be to take advantage of this improvement.
```
Comment (palpatim) [inline src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx:70]:
```suggestion
Only customers using the [Amplify GraphQL API CDK construct](https://constructs.dev/packages/@aws-amplify/graphql-api-construct/v/1.6.0?lang=typescript), will be able to take advantage of this improvement.
```

File: src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+### How did IAM authorization work before this improvement?
+
+There are two use cases for IAM auth with the Amplify GraphQL CDK construct. 
+
+**Use case 1: Provide a deny-by-default authorization pattern for Amazon Cognito identity pool's unauthenticated and authenticated role access to the API.** We wanted to ensure if you want any unauthenticated roles to access a model, you must explicit allow-list that behavior. For example: `@auth(rules: [{ allow: public, provider: iam }])`. 
+
+**Use case 2: Provide access to the GraphQL API from other AWS roles, such as AWS Lambda functions' execution role, or to use the GraphQL query explorer from the AWS AppSync console.** In this case, despite these IAM roles having the required IAM policies to access the GraphQL API, you needed to also supply these IAM roles into the Amplify GraphQL API CDK construct's `allowListedAdminRoles` properties. This often led to circular dependency issues that are difficult to untangle.
```
Comment (sobolk) [inline src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx:43]:
```suggestion
**Use case 2: Provide access to the GraphQL API from other IAM principals, such as AWS Lambda functions' execution role, or to use the GraphQL query explorer from the AWS AppSync console.** In this case, despite these IAM roles having the required IAM policies to access the GraphQL API, you needed to also supply these IAM roles into the Amplify GraphQL API CDK construct's `allowListedAdminRoles` properties. This often led to circular dependency issues that are difficult to untangle.
```

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx (hunk 1 above)
Comment (renebrandel) [reply src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:151]: After discussing on Slack, we'll do this later, when the new Gen 1 release is out.

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx (hunk 3 above)
Comment (renebrandel) [reply src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:699]: This section is specific to the Gen 1 CLI block of the docs, where they still have the old behaviro.

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx (hunk 6 above)
Comment (renebrandel) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:836]:
```suggestion
To grant any IAM principal (AWS Resource, IAM role, IAM user, etc), **with the exception of Amazon Cognito identity pool roles**, to this GraphQL API in CDK, you need to enable IAM authorization mode on the CDK construct.
```

File: src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+We'll issue a non-breaking feature improvement that'll make the use case clearer for customers.
+
+**Changes to use case 1:** To make the API more descriptive and clarify its intended behavior, we're renaming the `provider: iam` variable to `provider: identityPool` in the `@auth` rules. At the same time, we're introducing a `identityPoolConfig` property on the CDK construct that'll accept the existing values for (`identityPoolId`, `authenticatedUserRole`, `unauthenticatedUserRole`).
+
+**Changes to use case 2:** Instead of explicitly setting the `allowListedAuthRoles`, you now only need to set the `enableIamAuthorizationMode: true` under the `iamConfig` property. Now all resource access to the API with the exception of Amazon Cognito identity pool roles are governed by the resources' IAM role and policies.
+
+### How do I take advantage of this improvement?
+
+To take advantage of this improvement, follow these three step process:
```
Comment (renebrandel) [inline src/pages/[platform]/tools/cli/migration/iam-auth-updates-for-cdk-construct/index.mdx:55]:
```suggestion
To take advantage of this improvement, follow these steps:
```

**Review by sobolk — APPROVED** [review]

**Review by palpatim — APPROVED** [review]

**Review by sobolk — APPROVED** [review]

File: src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx [hunk 10]
```diff
@@ -170,7 +178,7 @@ new AmplifyGraphqlApi(this, "MyNewApi", {
     apiKeyConfig: {
       expires: cdk.Duration.days(30)
     },
-    iamConfig: {
+    identityPool: {
```
Comment (renebrandel) [inline src/pages/[platform]/build-a-backend/graphqlapi/customize-authorization-rules/index.mdx:181]:
```suggestion
    identityPoolConfig: {
```

**Review by palpatim — APPROVED** [review]

**Review by reesscot — APPROVED** [review]

**Review by swaminator — APPROVED** [review]

**Pull request merged**